lookup: mention `file` plugin exception (#74637)
ansible/ansible: docs/docsite/rst/plugins/lookup.rst

Lookup plugins are an Ansible-specific extension to the Jinja2 templating language. You can use lookup plugins to access data from outside sources (files, databases, key/value stores, APIs, and other services) within your playbooks. Like all :ref:`templating <playbooks_templating>`, lookups execute and are evaluated on the Ansible control machine. Ansible makes the data returned by a lookup plugin available using the standard templating system. You can use lookup plugins to load variables or templates with information from external sources.

.. note::
   - Pass ``wantlist=True`` to lookups to use in Jinja2 template "for" loops. See :ref:`query` for more details.

.. warning::
   - By default, lookup return values are marked as unsafe for security reasons. If you trust the outside source that your lookup accesses, pass ``allow_unsafe=True`` to allow Jinja2 templates to evaluate lookup values.
   - Some lookups pass arguments to a shell. When using variables from a remote/untrusted source, use the ``|quote`` filter to ensure safe usage.

.. _enabling_lookup:

Enabling lookup plugins
-----------------------

> Ansible enables all lookup plugins it can find. You do not need to maintain a list of enabled lookup plugins. To activate a custom lookup, place it in one of four locations:

Reviewer: this is not true anymore, it enables all those in 'configured paths' but for collections you have to do extra work

Reviewer 2: oh, interesting, so how do I enable a lookup plugin from a collection?

Reviewer: Simplest way, use fqcn and it will 'load on demand', though I believe we also added support with 'collections:' keywords to use short names, but that i need to confirm (nitz should know top of his head).

 Of course in both cases collection needs to be in collection search path.

Reviewer: some of my fixes will allow listing filters and plugins from collections also via `ansible-doc -l` but a bit stuck on that right now

Reviewer: Add note for collections:
```
 - If the lookup is part of a collection, the collection itself must be installed in the discoverable paths for collections (ref?)
 ```

> Ansible enables all lookup plugins it can find. You do not need to maintain a list of enabled lookup plugins. To activate a custom lookup, place it in one of four locations:

Reviewer 3: ```suggestion
Ansible enables all lookup plugins in the configured paths. You do not need to maintain a list of enabled lookup plugins. Use the fully-qualified collection name (FQCN) for lookup plugins in a collection and the collection itself must be installed in the discoverable paths for collections. See :ref:`collections` for details. To activate a custom lookup, place it in one of four locations:
```

> Ansible enables all lookup plugins it can find. You do not need to maintain a list of enabled lookup plugins. To activate a custom lookup, place it in one of four locations:

Reviewer 3: ```suggestion
Ansible enables all lookup plugins it can find. If the lookup is part of a collection, the collection itself must be installed in the :ref:`discoverable paths for collections <COLLECTIONS_PATHS>`. You do not need to maintain a list of enabled lookup plugins. To activate a custom lookup, place it in one of four locations:
```

.. note::
   - Lookups are executed with a working directory relative to the role or play,
     as opposed to local tasks, which are executed relative the executed script.

>      The ``file`` lookup plugin is exception to this as it has a list of path candidates in which it searches.

Reviewer: Those that look for files, vars or templates with a relative path will use the ``ansible_search_path``, to find those files.

Reviewer: probably with a link to https://docs.ansible.com/ansible/latest/user_guide/playbook_pathing.html

.. note::
   - Lookups are executed with a working directory relative to the role or play,
     as opposed to local tasks, which are executed relative the executed script.

>      The lookup plugins such as ``file``, ``fileglob``, ``template``, and ``filetree`` are exception to this as they have a list of path candidates to search files in.

Reviewer 3: ```suggestion
     The lookup plugins such as ``file``, ``fileglob``, ``template``, and ``filetree`` are exceptions to this as they have a list of path candidates to search files in.
```

>      The lookup plugins such as ``file``, ``fileglob``, ``template``, and ``filetree`` are exceptions to this as they have a list of path candidates to search files in.

Reviewer: they are not exceptions, they still execute relative to role or play

what they do differently is handle 'relative paths' in their search, but not in their execution

>      Also, note that lookup plugins that uses for ``files``, ``vars`` or ``templates`` parameters with a relative path will use the ``ansible_search_path``, to find those files.

Reviewer 3: ```suggestion
     Also, note that lookup plugins that use  ``files``, ``vars`` or ``templates`` parameters with a relative path use the ``ansible_search_path``, to find those files.
```

        - "{{ lookup('sequence', 'end=42 start=2 step=2')|map('log', 4)|list) }}"
        - ['a', 'c', 'd', 'c']

Understanding lookup plugin paths
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Lookups are executed with a working directory relative to the role or play (as opposed to local tasks, which are executed relative to the executed script). When searching for files and other resources, however, some lookups search additional paths.

Reviewer: we do not change working directory, this should read 'relative paths are calculated ...'also its not to play, bu to the file where the task is defined and/or role(s) it belongs to, link to the search paths docs as that is what this follows.

Reviewer: with that reference you should be able to skip most of what is below

Reviewer: Working directory does NOT change with role, only 'relative search paths' do and that is only for lookups that use the dwim functions.

Some lookups will make changes themselves, for example `pipe` changes the working directory to the 'playbook base dir'. But this is not something that happens universally and each lookup can do something different.
